pkg/auth/auth.go

```diff
@@ -0,0 +1,136 @@
+/*
+Package auth can be used for authentication and authorization
+Copyright 2018 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package auth
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	sdk_auth "github.com/libopenstorage/openstorage-sdk-auth/pkg/auth"
+)
+
+var (
+	requiredClaims = []string{"exp", "iat", "name", "email"}
+)
+
+type Authenticator interface {
+	AuthenticateToken(string) (*sdk_auth.Claims, error)
+}
+
+type JwtAuthConfig struct {
+	SharedSecret  []byte
+	RsaPublicPem  []byte
+	ECDSPublicPem []byte
+}
+
+type JwtAuthenticator struct {
+	config          JwtAuthConfig
+	rsaKey          interface{}
+	ecdsKey         interface{}
+	sharedSecretKey interface{}
+}
+
+func New(config *JwtAuthConfig) (*JwtAuthenticator, error) {
+
+	// Check at least one is set
+	if len(config.SharedSecret) == 0 &&
+		len(config.RsaPublicPem) == 0 &&
+		len(config.ECDSPublicPem) == 0 {
+		return nil, fmt.Errorf("Server was passed empty authentication information with no shared secret or pem files set")
+	}
+
+	authenticator := &JwtAuthenticator{
+		config: *config,
+	}
+
+	var err error
+	if len(config.SharedSecret) != 0 {
+		authenticator.sharedSecretKey = config.SharedSecret
+	}
+	if len(config.RsaPublicPem) != 0 {
+		authenticator.rsaKey, err = jwt.ParseRSAPublicKeyFromPEM(config.RsaPublicPem)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to parse rsa public key: %v", err)
+		}
+	}
+	if len(config.ECDSPublicPem) != 0 {
+		authenticator.ecdsKey, err = jwt.ParseECPublicKeyFromPEM(config.ECDSPublicPem)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to parse ecds public key: %v", err)
+		}
+	}
+
+	return authenticator, nil
+}
+
+func (j *JwtAuthenticator) AuthenticateToken(rawtoken string) (*sdk_auth.Claims, error) {
+
+	// Parse token
+	token, err := jwt.Parse(rawtoken, func(token *jwt.Token) (interface{}, error) {
```

Comment:
Is this where we check if the token has expired? If so, is `token.Valid` set to false?

If not, where do we check for token expiry? w/ `exp`

Reply:
Yeah, that's it. the jwt package sets that boolean value and the caller is supposed to check it.